src/levanter/trainer.py

```diff
@@ -549,18 +549,28 @@ def obj_fun(trainable_model):
         else:
             return loss, new_state, hook_infos
 
-    def _compute_gradients_microbatched(self, loss_fn, model: M, *batch, **batch_kwargs) -> tuple[Scalar, M]:
+    def _compute_gradients_microbatched(self, loss_fn, model: M, batch: X, **batch_kwargs) -> tuple[Scalar, M]:
         grad_fn = eqx.filter_value_and_grad(loss_fn, has_aux=False)
         mbs = self.config.microbatch_size
+        reduce = ReductionType.MEAN
+        if isinstance(batch, NumElementsBatch) and mbs != self.TrainBatch.size:
+            batch_kwargs[
+                "batch_num_elements"
+            ] = batch.num_elements()  # tell the loss function how many elements are in the batch
+            batch_kwargs[
+                "reduction"
+            ] = hax.sum  # the loss fn should sum the loss and divide by the number of elements, not average
+            reduce = ReductionType.SUM  # we're already normalizing the loss
         grad_fn = microbatched(
```

Comment:
let's have the grad_fn return both the total (masked) loss and the number of elements (and the gradient?). WDYT?

Reply:
So we'd change the signature of `ComputeLossFunction` to return something like
`(model, input) -> (loss, extras)` where `extras` has the number of elements in the microbatch in it?

I also find myself wanting to log things this way (by returning elements in an `extras` struct) so this would be a nice way to add that in.

The only concern is that if the number of microbatch elements or the losses are particularly large we could lose precision in the value of the loss right? 
When computing `(loss1 + loss2 + ... ) / num_elems` v.s. `loss1/num_elems + loss2/num_elems + ...`, I would kinda expect the second to be more accurate. Is this a reasonable concern or am I prematurely optimizing?

Reply:
Turns out the concern about precision is valid in certain regimes for fp16 but probably not for fp32

![image](https://github.com/user-attachments/assets/c237a0b3-2dc0-436e-b64b-3f4e1f16abb1)

I'll go ahead and try to work this more into the loss functions/grad accum!

Reply:
there's actually a standard way of dealing with this which I've encapsulated as [RunningMean](https://github.com/stanford-crfm/levanter/blob/07b3f1639ec4b2f05779733109a2906378e1af5a/src/levanter/utils/stat_utils.py#L13). The basic idea is to maintain the mean and the count rather than the sum and the count. This solves most precision problems usually. (You can actually do variance too but we don't need it)

Reply:
@dlwh I tried to encapsulate this in this last commit. I'm pretty sure it doesn't work right since it doesn't pass the `grad_accum` tests (trees don't match). Could you give it a quick peek and let me know if you see anything?